lightningd/peer_htlcs.c

```diff
@@ -779,14 +779,28 @@ htlc_accepted_hook_callback(struct htlc_accepted_hook_payload *request,
 	enum htlc_accepted_result result;
 	enum onion_type failure_code;
 	u8 *channel_update;
+	bool valid;
 	result = htlc_accepted_hook_deserialize(buffer, toks, &payment_preimage, &failure_code, &channel_update);
 
+	/* In order to handle it internally the payload needs to have at least
+	 * the `outgoing_cltv_value`, the `amt_to_forward` and the
+	 * `short_channel_id` if we are forwarding. */
+	valid = rs->amt_to_forward != NULL && rs->outgoing_cltv != NULL &&
+		(rs->nextcase == ONION_END ||
+		 (rs->nextcase == ONION_FORWARD && rs->forward_channel != NULL));
+
+	/* In addition we also enforce the TLV validity rules:
+	 *  - No unknown even types
+	 *  - Types in monotonical non-repeating order
+	 */
+	valid = valid && (rs->type == SPHINX_V0_PAYLOAD || tlv_payload_is_valid(rs->payload.tlv));
```

Comment:
seems like it'd be nicer to have a separate method for `validity check`, instead of making it all inline here.

Reply:
Added a `htlc_accepted_can_continue` function which tells us whether we can continue with the route-step if we're told to by the hook.